### PR TITLE
[SYCLomatic] Generate compile command for source files in linker entry

### PR DESCRIPTION
**Pull request by daiyaan-ahmed6**
[Fix]: Source files (.c, .cpp) are directly included in linker entries
through Makefiles in the compilation database. This fix generates a
compilation command for each source file in the linker entry into an
object file. The linker entry is generated seperately.

Signed-off-by: Ahmed, Daiyaan <daiyaan.ahmed@intel.com>

### CODE REVIEW

File: clang/test/dpct/compile_commands_linker_test/three.c [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// RUN: echo 0
+void foo(){
+    
+}
```
Comment (wenhuiNi) [inline clang/test/dpct/compile_commands_linker_test/three.c]: add new line

File: clang/test/dpct/compile_commands_linker_test/two.cu [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// RUN: echo 0
+__global__ void foo(){
+
+}
```
Comment (wenhuiNi) [inline clang/test/dpct/compile_commands_linker_test/two.cu]: add new line

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py [hunk 3]
```diff
@@ -269,6 +269,16 @@ def abspath(cwd, name):
         """Create normalized absolute path from input filename."""
         fullname = name if os.path.isabs(name) else os.path.join(cwd, name)
         return os.path.normpath(fullname)
+
```
Comment (wenhuiNi) [inline clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: When DPCT engineers  modifies the code of clang or llvm(not DPCT code), we need to add a macro # SYCLomatic_CUSTOMIZATION.
Please refer to https://github.com/oneapi-src/SYCLomatic/blob/6713571b76f528bf6761c12572e991319bba5d96/clang/tools/scan-build-py/lib/libscanbuild/arguments.py#L36C1-L36C1

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py [hunk 4]
```diff
@@ -282,13 +292,32 @@ def abspath(cwd, name):
...
             yield {
                 'directory': exec_trace['directory'],
                 'command': encode(command),
             }
+        # If linker entry contains source compilation files, generate output files for the source files, then generate linker entry
+        elif len(get_object_files(compilation.flags)) > 1 and compilation.files:
+            output_files = []
+            for source_idx in reversed(range(len(compilation.files))):
+                source = compilation.files[source_idx]
+                command = [compiler, "-c"] + filter_linker_entry(compilation.flags) + ['-o'] + [os.path.splitext(os.path.basename(source))[0]+'.o'] + [source]
```
Comment (wenhuiNi) [inline clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: need format

**Review by wenhuiNi — CHANGES_REQUESTED** [review]

File: clang/test/dpct/compile_commands_linker_test/three.c (hunk 1 above)
Comment (daiyaan-ahmed6) [reply clang/test/dpct/compile_commands_linker_test/three.c]: Added new line. Done

File: clang/test/dpct/compile_commands_linker_test/two.cu (hunk 2 above)
Comment (daiyaan-ahmed6) [reply clang/test/dpct/compile_commands_linker_test/two.cu]: Added new line. Done

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py (hunk 3 above)
Comment (daiyaan-ahmed6) [reply clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: Added macro as per reference.

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py (hunk 4 above)
Comment (daiyaan-ahmed6) [reply clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: Formatted using autopep8

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py [hunk 5]
```diff
@@ -290,6 +302,29 @@ def abspath(cwd, name):
                 'command': encode(command),
             }
 
+        # SYCLomatic_CUSTOMIZATION begin
```
Comment (wenhuiNi) [inline clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: ”# SYCLomatic_CUSTOMIZATION begin“ should be the start of this line. No space is need before it

Comment (daiyaan-ahmed6) [reply clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: Moved the macro to start of line

**Review by wenhuiNi — APPROVED** [review]

**Review by tangjj11 — APPROVED** [review]
lgtm

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py [hunk 6]
```diff
@@ -290,6 +302,29 @@ def abspath(cwd, name):
...
+# SYCLomatic_CUSTOMIZATION begin
+        # If linker entry contains source compilation files, generate output files for the source files, then generate linker entry
+        elif len(get_object_files(compilation.flags)) > 1 and compilation.files:
+            output_files = []
+            for source_idx in reversed(range(len(compilation.files))):
+                source = compilation.files[source_idx]
+                command = [compiler, "-c"] + filter_linker_entry(compilation.flags) + ['-o'] + [
+                    os.path.splitext(os.path.basename(source))[0]+'.o'] + [source]
+                output_files.append(os.path.splitext(
+                    os.path.basename(source))[0]+'.o')
```
Comment (tomflinda) [inline clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: Pls make sure the code style consistent with current script code style, for example there should be one space before and after operator "+" in Line314.

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py [hunk 7]
```diff
@@ -270,6 +270,17 @@ def abspath(cwd, name):
...
         return os.path.normpath(fullname)
 
+# SYCLomatic_CUSTOMIZATION begin
+    def get_object_files(flags):
+        return [flag for flag in flags if flag.endswith('.o')]
+
+    def filter_linker_entry(flags):
+        object_files = get_object_files(flags)
+        output_file_idx = flags.index('-o')
+        output_flags = flags[output_file_idx:output_file_idx+2]
```
Comment (tomflinda) [inline clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: Pls and comment for magic number "2", and also make sure the code style consistent with current script code style, for example there should be one space before and after operator "+".

File: clang/tools/scan-build-py/lib/libscanbuild/intercept.py (hunk 6 above)
Comment (daiyaan-ahmed6) [reply clang/tools/scan-build-py/lib/libscanbuild/intercept.py]: Fixed + operator to have consistent space.

**Review by tomflinda — APPROVED** [review]
LGTM

**Pull request merged**